File: kmsp11/fakekms/fakekms.go
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package fakekms contains a fake of the Google Cloud Key Management service.
// go/mocks#prefer-fakes
package fakekms

import (
	"net"
	"sync"
	"time"

	"cloud.google.com/kms/integrations/fakekms/fault"
	"google.golang.org/grpc"

	"cloud.google.com/go/kms/apiv1/kmspb"
	"cloud.google.com/kms/integrations/fakekms/fault/faultpb"
)

// maxPageSize is the maximum number of elements that will be returned in
// a single paged result of a list request.
const maxPageSize = 1000

// fakeKMS implements a fake of the Cloud KMS API.
type fakeKMS struct {
	kmspb.UnimplementedKeyManagementServiceServer
	keyRings map[keyRingName]*keyRing

	// Protects keyRings. For guarding object use within RPCs, the lock is held
	// in the lock interceptor rather than directly in the RPC function.
	mux sync.RWMutex
}

// keyRing models a KeyRing in Cloud KMS.
type keyRing struct {
	pb   *kmspb.KeyRing
	keys map[cryptoKeyName]*cryptoKey
}

// cryptoKey models a CryptoKey in Cloud KMS.
type cryptoKey struct {
	pb       *kmspb.CryptoKey
	versions map[cryptoKeyVersionName]*cryptoKeyVersion
}

func (f *fakeKMS) cryptoKey(name cryptoKeyName) (*cryptoKey, error) {
	kr, ok := f.keyRings[name.keyRingName]
	if !ok {
		return nil, errNotFound(name)
	}
	ck, ok := kr.keys[name]
	if !ok {
		return nil, errNotFound(name)
	}
	return ck, nil
}

// cryptoKeyVersion models a CryptoKeyVersion in Cloud KMS.
type cryptoKeyVersion struct {
	pb          *kmspb.CryptoKeyVersion
	keyMaterial interface{}
}

func (f *fakeKMS) cryptoKeyVersion(name cryptoKeyVersionName) (*cryptoKeyVersion, error) {
	kr, ok := f.keyRings[name.keyRingName]
	if !ok {
		return nil, errNotFound(name)
	}
	ck, ok := kr.keys[name.cryptoKeyName]
	if !ok {
		return nil, errNotFound(name)
	}
	ckv, ok := ck.versions[name]
	if !ok {
		return nil, errNotFound(name)
	}
	return ckv, nil
}

// Server wraps a local gRPC server that serves KMS requests.
type Server struct {
	Addr       net.Addr
	grpcServer *grpc.Server
}

// Close stops the server by immediately closing all connections and listeners.
func (s *Server) Close() {
	s.grpcServer.Stop()
}

// ServerOptions contains options for the FakeKMS server.
type ServerOptions struct {
	// The amount of time each request should be delayed before processing.
	Delay time.Duration
}

// NewServer starts a new local Fake KMS server that is listening for gRPC requests.
func NewServer() (*Server, error) {
	lis, err := net.Listen("tcp", "localhost:0")
	if err != nil {
		return nil, err
	}

	fakeKMS := &fakeKMS{keyRings: make(map[keyRingName]*keyRing)}
	faultServer := &fault.Server{}
	s := grpc.NewServer(grpc.ChainUnaryInterceptor(
		faultServer.NewInterceptor(), newLockInterceptor(&fakeKMS.mux)))
	kmspb.RegisterKeyManagementServiceServer(s, fakeKMS)
	faultpb.RegisterFaultServiceServer(s, faultServer)

	go s.Serve(lis)
	return &Server{Addr: lis.Addr(), grpcServer: s}, nil
}

File: kmsp11/fakekms/java/FakeKms.java
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.cloud.kms.pkcs11.fakekms;

import com.google.api.gax.core.NoCredentialsProvider;
import com.google.api.gax.grpc.GrpcTransportChannel;
import com.google.api.gax.rpc.FixedTransportChannelProvider;
import com.google.api.gax.rpc.TransportChannelProvider;
import com.google.cloud.kms.v1.KeyManagementServiceClient;
import com.google.cloud.kms.v1.KeyManagementServiceSettings;
import com.google.devtools.build.runfiles.Runfiles;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;

/** FakeKms provides a Java language binding to a fake Cloud KMS server. */
public class FakeKms implements AutoCloseable {

  private static final String FAKEKMS_PATH =
      "com_google_kmstools/fakekms/main/fakekms_/fakekms"
          + (System.getProperty("os.name").startsWith("Windows") ? ".exe" : "");

  private final Process process;
  private final String serverAddress;
  private ArrayList<GrpcTransportChannel> channels;

  /** Creates and starts a new Fake KMS server. */
  public FakeKms() throws IOException {
    String serverPath = Runfiles.create().rlocation(FAKEKMS_PATH);
    process = Runtime.getRuntime().exec(serverPath);
    serverAddress = new BufferedReader(new InputStreamReader(process.getInputStream())).readLine();
    channels = new ArrayList<>();
  }

  /** Returns a new KMS client that is wired to this fake. */
  public KeyManagementServiceClient newClient() throws IOException {
    GrpcTransportChannel channel = GrpcTransportChannel.create(
        ManagedChannelBuilder.forTarget(serverAddress).usePlaintext().build());
    channels.add(channel);
    TransportChannelProvider channelProvider =
        FixedTransportChannelProvider.create(channel);

    KeyManagementServiceSettings clientSettings =
        KeyManagementServiceSettings.newBuilder()
            .setTransportChannelProvider(channelProvider)
            .setCredentialsProvider(new NoCredentialsProvider())
            .build();

    return KeyManagementServiceClient.create(clientSettings);
  }

  public String getServerAddress() {
    return serverAddress;
  }

  /** Stops the fake server and releases all resources associated with it. */
  @Override
  public void close() {
    for (GrpcTransportChannel c : channels) {
      c.close();
    }

    if (process.isAlive()) {
      process.destroy();
    }
  }
}
